pkg/pki/ca/ca.go

```diff
@@ -80,6 +107,47 @@ func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string) (*IstioC
 // NewIstioCA returns a new IstioCA instance.
 func NewIstioCA(opts *IstioCAOptions) (*IstioCA, error) {
 	ca := &IstioCA{certTTL: opts.CertTTL}
+	// If the signing key/cert or root cert is empty, we should create a self-signed key/cert pair,
+	// and write it to the secret for persistent purpose.
+	// TODO(wattli): get rid of the NewSelfSignedIstioCA() after 0.2.
+	// K8s secret has a bug that you can't specify empty content, so we put a small content in it and
+	// use 10 as a threshold.
+	if len(opts.RootCertBytes) < 10 || len(opts.SigningCertBytes) < 10 || len(opts.SigningKeyBytes) < 10 {
+		now := time.Now()
+		options := CertOptions{
+			NotBefore:    now,
+			NotAfter:     now.Add(opts.CertTTL),
+			Org:          opts.Org,
+			IsCA:         true,
+			IsSelfSigned: true,
+			RSAKeySize:   caKeySize,
+		}
+		pemCert, pemKey := GenCert(options)
+		opts.CertChainBytes = []byte{}
+		opts.RootCertBytes = pemCert
+		opts.SigningCertBytes = pemCert
+		opts.SigningKeyBytes = pemKey
+
+		// Rewrite the key/cert back to secret so they will be persistent when CA restarts.
+		secret := &apiv1.Secret{
+			Data: map[string][]byte{
+				CACertChainID:  pemCert,
+				CAPrivateKeyID: pemKey,
+				RootCertID:     pemCert,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{serviceAccountNameAnnotationKey: CAServiceAccount},
+				Name:        SecretNamePrefix + CAServiceAccount,
+				Namespace:   opts.Namespace,
```

Comment:
This will like cause problem as Namespace could be empty.

Reply:
Fixed.